websocket/polygon.go

```diff
@@ -0,0 +1,228 @@
+					c.log.Errorf("connection closed unexpectedly: %v", err)
+					break
+				}
+				c.log.Errorf("failed to read message: %v", err)
+				break
+			}
+			c.rQueue <- msg
+		}
+	}
+}
+
+func (c *Client) write() {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		c.log.Debugf("closing write thread")
+		ticker.Stop()
+		c.conn.Close() // todo: should this force close?
+	}()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-ticker.C:
+			if err := c.conn.SetWriteDeadline(time.Now().Add(writeWait)); err != nil {
+				c.log.Errorf("failed to set write deadline: %v", err)
+				return // todo: should this return?
+			}
+			if err := c.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
+				c.log.Errorf("failed to send ping message: %v", err)
+				return
+			}
+		case msg := <-c.wQueue:
+			if err := c.conn.SetWriteDeadline(time.Now().Add(writeWait)); err != nil {
+				c.log.Errorf("failed to set write deadline: %v", err)
+				return // todo: should this return?
+			}
+			if err := c.conn.WriteMessage(websocket.TextMessage, msg); err != nil {
+				c.log.Errorf("failed to send message: %v", err)
+				return
+			}
+		}
+	}
+}
+
+func (c *Client) process() {
+	defer func() {
+		c.log.Debugf("closing process thread")
+	}()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case data := <-c.rQueue:
+			c.handle(data)
+		}
+	}
+}
+
+func (c *Client) handle(data []byte) {
+	var msgs []json.RawMessage
+	if err := json.Unmarshal(data, &msgs); err != nil {
+		c.log.Errorf("failed to process raw messages: %v", err)
+		return
+	}
+
+	for _, msg := range msgs {
+		if err := c.route(msg); err != nil { // todo: this might merit a "data router" type
+			c.log.Errorf("failed to process message: %v", err)
+		}
+	}
+}
+
+func (c *Client) route(msg json.RawMessage) error {
+	var ev models.EventType
+	err := json.Unmarshal(msg, &ev)
+	if err != nil {
+		return err
+	}
+
+	switch ev.EventType {
+	case "status":
+		return c.handleStatus(msg)
+	default:
+		c.log.Debugf("unknown message type '%v'", ev.EventType)
+	}
+
+	return nil
+}
+
+func (c *Client) handleStatus(msg json.RawMessage) error {
+	var cm models.ControlMessage
+	if err := json.Unmarshal(msg, &cm); err != nil {
+		return err
+	}
+
+	switch cm.Status {
+	case "connected":
+		c.log.Infof("connection successful")
+		b, err := json.Marshal(models.ControlMessage{
+			Action: "auth",
+			Params: c.apiKey,
+		})
+		if err != nil {
+			c.log.Errorf("authentication failed, closing connection")
+			c.Close() // fatal errors should close the connection
+		}
+		c.wQueue <- b
+	case "auth_success":
+		c.log.Infof("authentication successful")
+	case "auth_failed":
+		c.log.Errorf("authentication failed, closing connection")
+		c.Close()
+		return nil
+	case "success":
+		c.log.Infof("subscription successful") // todo: can subscriptions fail?
+	default:
+		c.log.Debugf("unknown status message '%v'", cm.Status)
+	}
```

Comment:
Curious about the log levels here. I would put the "* successful" messages as debug level and the "unknown status message" as info level. My thought being success is more or less expected, while an unknown status message might be something you'd want to see in production

Reply:
makes sense! i'll run through the existing code and clean things up a bit

Reply:
leaving things as is for now (it's helpful for developing) but i left a comment to clean things up as we get more stable